core/src/main/scala/org/broadinstitute/dsde/rawls/jobexec/SubmissionMonitorActor.scala

```diff
@@ -277,12 +278,26 @@ trait SubmissionMonitor extends FutureSupport with LazyLogging with RawlsInstrum
   private def execServiceOutputs(workflowRec: WorkflowRecord, petUser: UserInfo)(implicit executionContext: ExecutionContext): Future[Option[(WorkflowRecord, Option[ExecutionServiceOutputs])]] = {
     WorkflowStatuses.withName(workflowRec.status) match {
       case WorkflowStatuses.Succeeded =>
+        // asynchronously upload metadata to GCS.
+        // Currently, this is only used for hamm, and may potentially be removed from RAWLS if design changes
+        workflowRec.externalId.traverse{
+          workflowId =>
+            uploadMetadataToGCS(workflowRec.submissionId.toString, workflowId, petUser)
+        }
         executionServiceCluster.outputs(workflowRec, petUser).map(outputs => Option((workflowRec, Option(outputs))))
-
       case _ => Future.successful(Option((workflowRec, None)))
     }
   }
 
+  private def uploadMetadataToGCS(submissionId: String, workflowId: String, petUser: UserInfo)(implicit executionContext: ExecutionContext): Future[Unit] = {
+    for{
+      metadata <- executionServiceCluster.callLevelMetadataForCostCalculation(submissionId, workflowId, None, petUser)
+//    If this is a subworkflow, we don't upload its metadata; else we upload it to GCS
+      _ <- if(metadata.fields.keys.toList.contains("parentWorkflowId")) Future.successful(())
+      else googleServicesDAO.storeObject(GcsBucketName("cromwell_metadata"), GcsObjectName(workflowId), metadata.compactPrint.getBytes("UTF-8"))
```

Comment:
pull bucket name from config and there should be code to create the bucket on startup if it does not exist.

Reply:
> code to create the bucket on startup if it does not exist.

I wasn't sure if I should do that. What's the benefit of creating bucket in app?
For setup, we need to 1. create the bucket, 2. create a pubsub topic for this bucket and create notification. Are we doing all the setup in app?

Reply:
We should do as much setup in app as possible. This makes deployment much easier given all of our environments, fiab, etc. How we use terraform is not particularly suited this this at the moment specifically since it excludes fiab. But also it is not tied to our release process.